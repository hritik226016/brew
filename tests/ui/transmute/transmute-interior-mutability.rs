#![allow(unused)]
use std::{cell::UnsafeCell, mem, sync::atomic::AtomicI32};

fn main() {
    unsafe {
        mem::transmute::<&i32, &UnsafeCell<i32>>(&42);
        //~^ WARNING transmuting from a type without interior mutability to a type with interior mutability
        //~| HELP `UnsafeCell<i32>` has interior mutability
        // It's an error to transmute to a type containing unsafe cell
        mem::transmute::<&i32, &AtomicI32>(&42);
        //~^ WARNING transmuting from a type without interior mutability to a type with interior mutability
        //~| HELP `AtomicI32` has interior mutability
        // mutable_transmutes triggers before

        // This one is here because & -> &mut is worse, to assert that this one triggers.
        mem::transmute::<&i32, &mut UnsafeCell<i32>>(&42);
        //~^ ERROR transmuting &T to &mut T is undefined behavior, even if the reference is unused, consider instead using an UnsafeCell
    };
}
