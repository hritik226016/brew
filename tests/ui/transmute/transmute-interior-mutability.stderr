warning: transmuting from a type without interior mutability to a type with interior mutability is undefined behaviour if you modify the content
  --> $DIR/transmute-interior-mutability.rs:6:9
   |
LL |         mem::transmute::<&i32, &UnsafeCell<i32>>(&42);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: `UnsafeCell<i32>` has interior mutability
   = note: `#[warn(transmute_to_interior_mutability)]` on by default

warning: transmuting from a type without interior mutability to a type with interior mutability is undefined behaviour if you modify the content
  --> $DIR/transmute-interior-mutability.rs:10:9
   |
LL |         mem::transmute::<&i32, &AtomicI32>(&42);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = help: `AtomicI32` has interior mutability

error: transmuting &T to &mut T is undefined behavior, even if the reference is unused, consider instead using an UnsafeCell
  --> $DIR/transmute-interior-mutability.rs:16:9
   |
LL |         mem::transmute::<&i32, &mut UnsafeCell<i32>>(&42);
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: `#[deny(mutable_transmutes)]` on by default

error: aborting due to 1 previous error; 2 warnings emitted

